### ISSUE TITLE
'use client' directive is not working when module type is 'umd'

### ISSUE THREAD

**Issue by sohnjunior** [swc-project/swc#8728]
### Describe the bug

```typescript
// input.ts

'use client';

export function ClientComponent() {
  return 'Hello world';
}
```

```bash
> swc input.ts -o output.js -C module.type=umd
```

```javascript
// output.js

(function(global, factory) {
    if (typeof module === "object" && typeof module.exports === "object") factory(exports);
    else if (typeof define === "function" && define.amd) define([
        "exports"
    ], factory);
    else if (global = typeof globalThis !== "undefined" ? globalThis : global || self) factory(global.input = {});
})(this, function(exports) {
    "use client";
    "use strict";
    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    Object.defineProperty(exports, "ClientComponent", {
        enumerable: true,
        get: function() {
            return ClientComponent;
        }
    });
    function ClientComponent() {
        return "Hello world";
    }
});


//# sourceMappingURL=output.js.map

```

The 'use client' directive is defined inside the factory function.

Perhaps because of this, an error saying 'use client directive does not exist' is occurring in the Next.js environment that uses the App directory.

This issue is related with [#7315](https://github.com/swc-project/swc/issues/7315)

### Playground link (or link to the minimal reproduction)

https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAAx3LMQqAMAxG4T2n%2BLfqYg%2Fg2MV71AiFmpQ2RUG8u8XtwcfzHklKt8Uaud4YMScWcysR30Wr4egSLakg%2FBL0LCojphkPAZWtV4HbOGfFpTXvY37pA1C%2FhblZAAAA&config=H4sIAAAAAAAAA1WPSw7DIAwF95wCed1tN71DD4GIE1GFj2wjFUW5eyGFtNnh98YMbEpreLGFh97qsQ7JECOdc024BDHvmoCUhGzJJYHbaIVbNZuV8Yj2bwNiaEFpW8j3jsMaI%2BPAe%2BZdcHP5F9roEyHzFWyoCcuKV53qSvBxykfZP9Ie2%2FTZT%2FCDhuy8GBw%2Fx6ZQRrV%2FAL%2FpEOoUAQAA

### Expected behavior

The 'use client' directive is kept at the top of the file.

### Version

1.3.101

**Comment by magic-akari** [comment]
> The `use strict` directive must be inside the UMD function and not outside, because UMD is safe to be used with simple file concatenation and thus directives at the top of the files would risk getting lost (and just transformed into normal string expression statements)
>
> from @nicolo-ribaudo (https://x.com/NicoloRibaudo/status/1767871497400373272)

I fully agree with this statement. For UMD (as well as similar AMD and SystemJS), the actual module code resides within its encapsulated portion. Given that the toolchain needs to accommodate more generic logic, we aim not to disrupt the existing logic.

In the case of UMD, there is a common expectation that it can be easily concatenated together. In such scenarios, we desire directives to solely impact the code within the module, rather than the global code.

The root of this issue lies in the differences between toolchain's expectations of directives and how React perceives directives. Within React, directives operate at the file level and must be at the top of the file to take effect. Conversely, in the toolchain's perspective, directives function at the block level and can be safely relocated to their new scope during the transformation process.

This misalignment of expectations has led to several challenges, prompting various approaches to address the issue:
1. The first involves starting from React, enabling React to recognize UMD/AMD/SystemJS module code. However, due to the flexibility of module wrappers, this approach proves challenging in practice.
2. The second approach highlights the incongruity between using UMD and React's expectations for client components, where client components cannot be freely concatenated. Therefore, opting for alternative module systems such as CJS or sticking with ESM might be more suitable.
3. The final approach is crafting a custom SWC plugin to tackle this scenario. You can lift directives within SWC plugins.

**Comment by sohnjunior** [comment]
Thank you for your feedback. Fortunately, I decided to adopt the CJS module method instead of the UMD method.
